lms/djangoapps/program_enrollments/signals.py

```diff
@@ -0,0 +1,47 @@
+"""
+Signal handlers for program enrollments
+"""
+import logging
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+from social_django.models import UserSocialAuth
+from student.models import CourseEnrollmentException
+from third_party_auth.models import SAMLProviderConfig
+
+from lms.djangoapps.program_enrollments.models import ProgramEnrollment
+
+logger = logging.getLogger(__name__)
+
+
+@receiver(post_save, sender=UserSocialAuth)
+def populate_waiting_enrollments(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Post-save signal to update any waiting program enrollments with a user,
+    and enroll the user in any waiting course enrollments.
+
+    In most cases this will just short-circuit. Enrollments will only be updated
+    for a SAML provider with a linked organization.
+    """
+    user = instance.user
+
+    try:
+        provider_slug, external_user_key = instance.uid.split(':')
+        if not SAMLProviderConfig.objects.get(slug=provider_slug).organization:
+            return
+    except (SAMLProviderConfig.DoesNotExist, ValueError):
+        return
+
+    incomplete_enrollments = ProgramEnrollment.objects.filter(
+        external_user_key=external_user_key
+    ).prefetch_related('program_course_enrollments')
+
+    incomplete_enrollments.update(user=user)
+
+    for enrollment in incomplete_enrollments:
+        for program_course_enrollment in enrollment.program_course_enrollments.all():
+            try:
+                program_course_enrollment.enroll(user)
+            except CourseEnrollmentException as e:
+                logger.critical(
```

Comment:
I'd agree that if we're causing the social_auth creation to fail on an exception we don't need critical here.  But are we certain we want the user totally locked out if one course has a problem?

As far as trying to enroll() on an existing enrollment.  This will throw a CONFLICT exception that we'll need to ignore.  I can easily add a test case for that. However, don't we need to update the existing enrollment if the user is `audit` track?

Reply:
I don't think the learner will get locked out. They most likely will retry the whole SSO process again if they see such an error, and retry would fail. They can still log into their edx account. They can still log into their school account. They just don't see them getting linked. Which means the SSO would not succeed.
As for `audit` track, that is a good product question. Can you hash this case out with Deen? My opinion is that, we don't touch existing enrollments for now because we don't have full control over the impact if we do. But the best way is to hash this case out with Deen.

Reply:
makes sense to me.  I'll catch up with Deen on `audit`

Reply:
I don't think this is a critical error. What if the learner, before linked to the school, already got enrolled and even completed in some Masters program courses?
Also, a question is, what if those existing enrollment has the learner in `verified` track. Do you think this step will move those learners into `Masters` track? I think the correct behavior is that, if the learner is already enrolled, we don't touch the enrollment. Can you please test for this case in unit tests as well?